src/platform/test/mtest-main.c

```diff
@@ -0,0 +1,202 @@
+
+#ifdef M_CORE_GBA
+	if (core->platform(core) == mPLATFORM_GBA) {
+		((struct GBA*) core->board)->hardCrash = false;
+
+		_exitSwiImmediate = mTestOpts.exitSwiImmediate;
+		_returnCodeRegister = mTestOpts.returnCodeRegister;
+
+		if (_exitSwiImmediate == 3) {
+			// Hook into SWI 3 (shutdown)
+			struct mCoreCallbacks callbacks = {0};
+			callbacks.context = core;
+			callbacks.shutdown = (void(*)(void*)) _mTestSwi3Callback;
+			core->addCoreCallbacks(core, &callbacks);
+		} else {
+			// Custom SWI hooks
+			_armSwi16 = ((struct GBA*) core->board)->cpu->irqh.swi16;
+			((struct GBA*) core->board)->cpu->irqh.swi16 = _mTestSwi16;
+			_armSwi32 = ((struct GBA*) core->board)->cpu->irqh.swi32;
+			((struct GBA*) core->board)->cpu->irqh.swi32 = _mTestSwi32;
+		}
+	}
+#endif
+
+	bool cleanExit = true;
+	if (!mCoreLoadFile(core, args.fname)) {
+		cleanExit = false;
+		goto loadError;
+	}
+	if (args.patch) {
+		core->loadPatch(core, VFileOpen(args.patch, O_RDONLY));
+	}
+
+	struct VFile* savestate = NULL;
+
+	if (args.savestate) {
+		savestate = VFileOpen(args.savestate, O_RDONLY);
+	}
+
+	core->reset(core);
+
+	struct mCheatDevice* device;
+	if (args.cheatsFile && (device = core->cheatDevice(core))) {
+		struct VFile* vf = VFileOpen(args.cheatsFile, O_RDONLY);
+		if (vf) {
+			mCheatDeviceClear(device);
+			mCheatParseFile(device, vf);
+			vf->close(vf);
+		}
+	}
+
+	if (savestate) {
+		mCoreLoadStateNamed(core, savestate, 0);
+		savestate->close(savestate);
+	}
+
+	do {
+		core->runLoop(core);
+	} while (!_dispatchExiting);
+
+	core->unloadROM(core);
+
+loadError:
+	freeArguments(&args);
+	mCoreConfigDeinit(&core->config);
+	core->deinit(core);
+
+	return cleanExit ? _exitCode : 1;
+}
+
+static void _mTestShutdown(int signal) {
+	UNUSED(signal);
+	_dispatchExiting = true;
+}
+
+#ifdef M_CORE_GBA
+static void _mTestSwi3Callback(struct mCore* core) {
+	_exitCode = ((struct GBA*) core->board)->cpu->regs.gprs[_returnCodeRegister];
+	_dispatchExiting = true;
+}
+
+static void _mTestSwi16(struct ARMCore* cpu, int immediate) {
+	if (immediate == _exitSwiImmediate) {
+		_exitCode = cpu->regs.gprs[_returnCodeRegister];
+		_dispatchExiting = true;
+		return;
+	}
+	_armSwi16(cpu, immediate);
+}
+
+static void _mTestSwi32(struct ARMCore* cpu, int immediate) {
+	if (immediate == _exitSwiImmediate) {
+		_exitCode = cpu->regs.gprs[_returnCodeRegister];
+		_dispatchExiting = true;
+		return;
+	}
+	_armSwi32(cpu, immediate);
+}
+#endif
+
+static bool _parseMTestOpts(struct mSubParser* parser, int option, const char* arg) {
+	struct MTestOpts* opts = parser->opts;
+	errno = 0;
+	switch (option) {
+	case 'S':
+		opts->exitSwiImmediate = strtol(arg, NULL, 0);
+		return !errno;
+	case 'R':
+		opts->returnCodeRegister = _parseNamedRegister(arg);
+		return !errno;
+	default:
+		return false;
+	}
+}
+
+static int _parseNamedRegister(const char* regStr) {
+	if (regStr[0] == 'r' || regStr[0] == 'R') {
+		++regStr;
+	}
+	return strtol(regStr, NULL, 10);
```

Comment:
It would probably make sense to properly constrain this. Not that I'm too worried about some malicious actor here, but it is an obvious place to do so. Some error checking that it didn't fail to parse would probably be good too.